kernel/src/syscall.rs

```diff
@@ -522,111 +552,197 @@ impl SyscallReturn {
     /// Encode the system call return value into 4 registers, following the
     /// encoding specified in TRD104. Architectures which do not follow TRD104
     /// are free to define their own encoding.
+    /// TODO: deprecate in favour of the more general one
     pub fn encode_syscall_return(&self, a0: &mut u32, a1: &mut u32, a2: &mut u32, a3: &mut u32) {
+        if core::mem::size_of::<MetaPtr>() == core::mem::size_of::<u32>() {
+            // SAFETY: if the two unsized integers are the same size references to them
+            // can be safely transmuted.
+            // Ugly coercion could be avoided by first copying to the stack, then assigning with
+            // "as" in order to satisfy the compiler. But I expect this function will disappear
+            // in favour of just using the usize one.
+            unsafe {
+                let a0 = &mut *(core::ptr::from_mut(a0) as *mut MetaPtr);
+                let a1 = &mut *(core::ptr::from_mut(a1) as *mut MetaPtr);
+                let a2 = &mut *(core::ptr::from_mut(a2) as *mut MetaPtr);
+                let a3 = &mut *(core::ptr::from_mut(a3) as *mut MetaPtr);
+                self.encode_syscall_return_mptr(a0, a1, a2, a3);
+            }
+        } else {
+            panic!("encode_syscall_return used on a 64-bit platform or CHERI platform")
+        }
+    }
+
+    /// The obvious extension of TRD104 that works for 32-bit and 64-bit platforms.
+    /// This makes no changes to TRD104 on 32-bit platforms.
```

Comment:
Eventually I think we will benefit from a dedicated PR / discussion thread updating TRD104 and more explicitly discussing the best ABI for capabilities.

I'm not sure we should necessarily block this PR on that longer process.. but this code does change our ABI, which feels like it should be partitioned off somehow more explicitly from anything not using CHERI in the short term; I'm not sure the least-painful way to do that :/

Reply:
I am doing to take a stab at TRD104 now